### PR TITLE
Element penalty calculation for MFEM interface

### LINKED ISSUE #17: Add element-based penalty to MFEM interface
Let's decide on an interface for the element-based penalty for the MFEM interface.  Should the penalty be calculated in tribol or the host code?

### PR DESCRIPTION
**Pull request by ebchin**
Initial implementation of element penalty calculation for the MFEM interface.

### CODE REVIEW

File: src/tribol/interface/mfem_tribol.hpp [hunk 1]
```diff
@@ -91,6 +91,56 @@ void registerMfemCouplingScheme( integer cs_id,
...
+ * registerMfemCouplingScheme()
+ *
+ * @param cs_id The ID of the coupling scheme with the MFEM mesh
+ * @param mesh1_penalty Penalty parameter for the first contact surface mesh
+ * @param mesh2_penalty Penalty parameter for the second contact surface mesh
+ */
+void setMfemConstantPenalty( integer cs_id, real mesh1_penalty, real mesh2_penalty );
+
+/**
+ * @brief Computes element thickness for the volume elements associated with the
```
Comment (srwopschall) [inline src/tribol/interface/mfem_tribol.hpp]: Maybe include that this is the element thickness at the face centroid?

File: src/examples/mfem_common_plane.cpp [hunk 2]
```diff
@@ -352,6 +352,8 @@ int main( int argc, char** argv )
     tribol::KINEMATIC_CONSTANT,
     tribol::NO_RATE_PENALTY
   );
+  // This set the penalty parameters on the contact meshes.
+  tribol::setMfemConstantPenalty(coupling_scheme_id, p_kine, p_kine);
```
Comment (srwopschall) [inline src/examples/mfem_common_plane.cpp]: Can we keep names consistent with c-style API? So this would be setMfemKinematicConstantPenalty()? I like the clarity in distinguishing what is being penalized (kinematic gap or rate gap).

File: src/tribol/interface/mfem_tribol.hpp [hunk 3]
```diff
@@ -91,6 +91,108 @@ void registerMfemCouplingScheme( integer cs_id,
...
+ *
+ * @pre Coupling scheme cs_id must be registered using registerMfemCouplingScheme()
+ * @pre Kinematic penalty must be initialized through setMfemKinematicConstantPenalty() or
+ * setMfemKinematicElementPenalty()
+ * 
+ * @param cs_id The ID of the coupling scheme with the MFEM mesh
+ * @param mesh1_scale Scaling coefficient of the kinematic penalty for the first contact surface mesh
+ * @param mesh2_scale Scaling coefficient of the kinematic penalty for the second contact surface mesh
+ */
+void setMfemRatePercentPenalty( integer cs_id, real mesh1_scale, real mesh2_scale );
```
Comment (srwopschall) [inline src/tribol/interface/mfem_tribol.hpp]: Can we call these '_ratio'?

File: src/tribol/interface/mfem_tribol.hpp [hunk 4]
```diff
@@ -91,6 +91,108 @@ void registerMfemCouplingScheme( integer cs_id,
...
+ * 
+ * @pre Coupling scheme cs_id must be registered using registerMfemCouplingScheme()
+ * @pre Kinematic penalty must be initialized through setMfemKinematicConstantPenalty() or
+ * setMfemKinematicElementPenalty()
+ *
+ * @param cs_id The ID of the coupling scheme with the MFEM mesh
+ * @param mesh1_scale Scaling coefficient of the kinematic penalty for the first contact surface mesh
+ * @param mesh2_scale Scaling coefficient of the kinematic penalty for the second contact surface mesh
+ */
+void setMfemPenaltyScale( integer cs_id, real mesh1_scale, real mesh2_scale );
```
Comment (srwopschall) [inline src/tribol/interface/mfem_tribol.hpp]: Change to setMfemKinematicPenaltyScale()

File: src/tribol/interface/mfem_tribol.hpp [hunk 5]
```diff
@@ -91,6 +91,108 @@ void registerMfemCouplingScheme( integer cs_id,
...
+
+/**
+ * @brief Computes element thickness for the volume elements associated with the contact surface mesh.
+ *
+ * Element thickness is calculated at the origin of the isoparametric volume element in the direction given by the
+ * normal at the center of the associated isoparametric surface element.
```
Comment (srwopschall) [inline src/tribol/interface/mfem_tribol.hpp:164]: Add note for use case.

File: src/tribol/interface/mfem_tribol.cpp [hunk 6]
```diff
@@ -119,6 +118,133 @@ void setMfemLORFactor( integer cs_id,
...
+      "Create the coupling scheme using registerMfemCouplingScheme() to set the penalty."
+   );
+   auto penalty_opts = coupling_scheme->getEnforcementOptions().penalty_options;
+   SLIC_ERROR_ROOT_IF(
+      !penalty_opts.kinematic_calc_set,
+      "No kinematic enforcement method set. Call setMfemKinematicConstantPenalty() or "
+      "setMfemKinematicElementPenalty() first."
+   );
+   setPenaltyOptions(
+      cs_id, penalty_opts.constraint_type, penalty_opts.kinematic_calculation, RATE_CONSTANT);
```
Comment (srwopschall) [inline src/tribol/interface/mfem_tribol.cpp]: penalty_opts.constraint_type should be updated/changed to KINEMATIC_AND_RATE (or whatever the enum is).

File: src/tribol/interface/mfem_tribol.cpp [hunk 7]
```diff
@@ -119,6 +118,133 @@ void setMfemLORFactor( integer cs_id,
...
+      "Create the coupling scheme using registerMfemCouplingScheme() to set the penalty."
+   );
+   auto penalty_opts = coupling_scheme->getEnforcementOptions().penalty_options;
+   SLIC_ERROR_ROOT_IF(
+      !penalty_opts.kinematic_calc_set,
+      "No kinematic enforcement method set. Call setMfemKinematicConstantPenalty() or "
+      "setMfemKinematicElementPenalty() first."
+   );
+   setPenaltyOptions(
+      cs_id, penalty_opts.constraint_type, penalty_opts.kinematic_calculation, RATE_PERCENT);
```
Comment (srwopschall) [inline src/tribol/interface/mfem_tribol.cpp]: penalty_opts.constraint_type should be updated to KINEMATIC_AND_RATE.

File: src/tribol/interface/mfem_tribol.cpp [hunk 8]
```diff
@@ -119,6 +118,133 @@ void setMfemLORFactor( integer cs_id,
...
+{
+   auto coupling_scheme = CouplingSchemeManager::getInstance().getCoupling(cs_id);
+   SLIC_ERROR_ROOT_IF(
+      !coupling_scheme->hasMfemData(),
+      "Coupling scheme does not contain MFEM data. "
+      "Create the coupling scheme using registerMfemCouplingScheme() to set the penalty."
+   );
+   auto penalty_opts = coupling_scheme->getEnforcementOptions().penalty_options;
+   SLIC_ERROR_ROOT_IF(
+      !penalty_opts.kinematic_calc_set,
```
Comment (srwopschall) [inline src/tribol/interface/mfem_tribol.cpp:210]: Double check header file for a \pre or \note explaining that the kinematic penalty has to be set first.

File: src/tribol/interface/mfem_tribol.cpp [hunk 9]
```diff
@@ -269,6 +395,69 @@ void updateMfemParallelDecomposition()
...
+               setRateConstantPenalty(mesh_ids[1], *mfem_data->GetMesh2RatePenalty());
+            }
+            else if (penalty_opts.rate_calculation == RATE_PERCENT)
+            {
+               SLIC_ERROR_ROOT_IF(
+                  !mfem_data->GetMesh1Penalty() || !mfem_data->GetMesh2Penalty(),
+                  "Penalty parameters have not been set.  Call setMfemConstantPenalty()."
+               );
+               setRatePercentPenalty(mesh_ids[0], *mfem_data->GetMesh1RateScale());
+               setRatePercentPenalty(mesh_ids[0], *mfem_data->GetMesh2RateScale());
```
Comment (srwopschall) [inline src/tribol/interface/mfem_tribol.cpp]: Update mesh_ids[1] for second mesh id.

File: src/tribol/mesh/MfemData.hpp [hunk 10]
```diff
@@ -810,6 +810,200 @@ class MfemMeshData
...
+
+  /**
+   * @brief Clears rate penalty data
+   */
+  void ClearRatePenaltyData();
+
+  /**
+   * @brief Sets the penalty parameter for the first registered Tribol mesh
+   * 
+   * @param penalty Penalty value for the first registered Tribol mesh
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp:838]: Mention that this is a single constant penalty.

File: src/tribol/mesh/MfemData.hpp [hunk 11]
```diff
@@ -810,6 +810,200 @@ class MfemMeshData
...
+   */
+  void SetMesh1Penalty(real penalty) { penalty_1_ = std::make_unique<double>(penalty); }
+
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp:844]: Change name to reflect that you are setting the Kinematic Constant penalty.

File: src/tribol/mesh/MfemData.hpp [hunk 12]
```diff
@@ -810,6 +810,200 @@ class MfemMeshData
...
+  /**
+   * @brief Sets the penalty parameter for the second registered Tribol mesh
+   * 
+   * @param penalty Penalty value for the second registered Tribol mesh
+   */
+  void SetMesh2Penalty(real penalty) { penalty_2_ = std::make_unique<double>(penalty); }
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp]: See notes above for Mesh1

File: src/tribol/mesh/MfemData.hpp [hunk 13]
```diff
@@ -810,6 +810,200 @@ class MfemMeshData
...
+
+  /**
+   * @brief Get the penalty parameter for the first registered Tribol mesh
+   *
+   * @return const real* 
+   */
+  const real* GetMesh1Penalty() const { return penalty_1_.get(); }
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp]: Update name for all kinematic constant related routines.

File: src/tribol/mesh/MfemData.hpp [hunk 14]
```diff
@@ -810,6 +810,200 @@ class MfemMeshData
...
+   * @return const real* 
+   */
+  const real* GetMesh2PenaltyScale() const { return scale_2_.get(); }
+
+  /**
+   * @brief Sets the rate penalty for the first registered Tribol mesh
+   * 
+   * @param penalty Rate penalty value for the first registered Tribol mesh
+   */
+  void SetMesh1RatePenalty(real penalty) { rate_penalty_1_ = std::make_unique<double>(penalty); }
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp]: Update 'Rate' getters/setters to reflect whether it is rate constant or rate percent/ratio

File: src/tribol/mesh/MfemData.hpp [hunk 15]
```diff
@@ -810,6 +810,200 @@ class MfemMeshData
...
+   * @return const real* 
+   */
+  const real* GetMesh2RatePenalty() const { return rate_penalty_2_.get(); }
+
+  /**
+   * @brief Sets the rate penalty scale for the first registered Tribol mesh
+   * 
+   * @param scale Rate penalty scale value for the first registered Tribol mesh
+   */
+  void SetMesh1RateScale(real scale) { rate_scale_1_ = std::make_unique<double>(scale); }
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp]: Change name to Percent for consistency.

File: src/tribol/mesh/MfemData.hpp [hunk 16]
```diff
@@ -810,6 +810,200 @@ class MfemMeshData
...
+
+  /**
+   * @brief Sets the rate penalty scale for the second registered Tribol mesh
+   * 
+   * @param scale Rate penalty scale value for the second registered Tribol mesh
+   */
+  void SetMesh2RateScale(real scale) { rate_scale_2_ = std::make_unique<double>(scale); }
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp]: See note above.

File: src/tribol/mesh/MfemData.hpp [hunk 17]
```diff
@@ -810,6 +810,200 @@ class MfemMeshData
...
+   * @return const real* 
+   */
+  const real* GetMesh1RateScale() const { return rate_scale_1_.get(); }
+
+  /**
+   * @brief Get the rate penalty scale for the second registered Tribol mesh
+   *
+   * @return const real* 
+   */
+  const real* GetMesh2RateScale() const { return rate_scale_2_.get(); }
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp]: See 'percent' note above.

File: src/tribol/mesh/MfemData.hpp [hunk 18]
```diff
@@ -810,6 +810,200 @@ class MfemMeshData
...
+   * @brief Get a pointer to the element thickness array for the first Tribol
+   * registered mesh
+   *
+   * @return const real* 
+   */
+  const real* GetRedecompElemThickness1() const
+  {
+    if (tribol_elem_thickness_1_)
+    {
+      return &(*tribol_elem_thickness_1_)[0];
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp]: Clean this line up.

File: src/tribol/mesh/MfemData.hpp [hunk 19]
```diff
@@ -810,6 +810,200 @@ class MfemMeshData
...
+   */
+  const real* GetMesh2RateScale() const { return rate_scale_2_.get(); }
+
+  /**
+   * @brief Get a pointer to the element thickness array for the first Tribol
+   * registered mesh
+   *
+   * @return const real* 
+   */
+  const real* GetRedecompElemThickness1() const
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp:973]: Double check that this routine/check works with zero element redecomp meshes. There were API calls that check if this is a non-null pointer, but looks like it will be for zero element redecomp meshes which will kill a simulation (error out).

File: src/tribol/mesh/MfemData.hpp [hunk 20]
```diff
@@ -810,6 +810,200 @@ class MfemMeshData
...
+    }
+  }
+
+  /**
+   * @brief Get a pointer to the material modulus array for the first Tribol
+   * registered mesh
+   *
+   * @return const real* 
+   */
+  const real* GetRedecompMaterialModulus1() const
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp:995]: See notes above about GetRedecompElemThickness<1,2>()

File: src/tribol/mesh/MfemData.hpp [hunk 21]
```diff
@@ -904,6 +1098,15 @@ class MfemMeshData
...
    */
   void SetLORFactor(integer lor_factor);
+
+  /**
+   * @brief Computes element thicknesses for volume elements attached to the
+   * contact surface
+   */
+  void ComputeElementThicknesses();
+
+  void SetMaterialModulus(mfem::Coefficient& modulus_field);
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp:1132]: Add doxygen

File: src/tribol/mesh/MfemData.hpp [hunk 22]
```diff
@@ -1106,6 +1309,88 @@ class MfemMeshData
    */
   std::unique_ptr<ParentField> velocity_;
 
+  /**
+   * @brief Contact penalty for the first Tribol registered mesh
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp]: Mention that this is a the single constant value.

File: src/tribol/mesh/MfemData.hpp [hunk 23]
```diff
@@ -1106,6 +1309,88 @@ class MfemMeshData
...
+   */
+  std::unique_ptr<real> penalty_1_;
+
+  /**
+   * @brief Contact penalty for the second Tribol registered mesh
+   */
+  std::unique_ptr<real> penalty_2_;
+
+  /**
+   * @brief Contact penalty scale for the first Tribol registered mesh
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp]: Mention that this scales kinematic penalty

File: src/tribol/mesh/MfemData.hpp [hunk 24]
```diff
@@ -1106,6 +1309,88 @@ class MfemMeshData
...
+   */
+  std::unique_ptr<real> scale_1_;
+
+  /**
+   * @brief Contact penalty scale for the second Tribol registered mesh
+   */
+  std::unique_ptr<real> scale_2_;
+
+  /**
+   * @brief Contact rate penalty for the first Tribol registered mesh
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp]: Add what kind of rate

File: src/tribol/mesh/MfemData.hpp [hunk 25]
```diff
@@ -1106,6 +1309,88 @@ class MfemMeshData
...
+
+  /**
+   * @brief Contact rate penalty for the second Tribol registered mesh
+   */
+  std::unique_ptr<real> rate_penalty_2_;
+
+  /**
+   * @brief Contact rate penalty scale for the first Tribol registered mesh
+   */
+  std::unique_ptr<real> rate_scale_1_;
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.hpp]: Change to rate_percent for consistency

File: src/tribol/mesh/MfemData.cpp [hunk 26]
```diff
@@ -471,6 +553,98 @@ void MfemMeshData::SetLORFactor(integer lor_factor)
...
+    auto parent_bdr_e = submesh_.GetParentElementIDMap()[submesh_e];
+    auto& parent_mesh = const_cast<mfem::ParMesh&>(parent_mesh_);
+    auto& face_el_tr = *parent_mesh.GetBdrFaceTransformations(parent_bdr_e);
+    auto mask = face_el_tr.GetConfigurationMask();
+    auto parent_e = (mask & mfem::FaceElementTransformations::HAVE_ELEM1) ?
+      face_el_tr.Elem1No :
+      face_el_tr.Elem2No;
+    
+    // Step 2
+    // norm = (dx/dxi x dx/deta) / || dx/dxi x dx/deta || on boundary element
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.cpp]: call 'normal'. Distinguish that the boundary element here is the parent volume boundary element.

File: src/tribol/mesh/MfemData.cpp [hunk 27]
```diff
@@ -471,6 +553,98 @@ void MfemMeshData::SetLORFactor(integer lor_factor)
...
+    mfem::DenseMatrix elem_coords(parent_mesh_.Dimension(), be_dofs.Size());
+    for (int d{0}; d < parent_mesh_.Dimension(); ++d)
+    {
+      mfem::Array<int> be_vdofs(be_dofs);
+      parent_fes.DofsToVDofs(d, be_vdofs);
+      mfem::Vector elemvect(be_dofs.Size());
+      coords_.GetParentGridFn().GetSubVector(be_vdofs, elemvect);
+      elem_coords.SetRow(d, elemvect);
+    }
+    auto& be = *parent_fes.GetBE(parent_bdr_e);
```
Comment (srwopschall) [inline src/tribol/mesh/MfemData.cpp:618]: Maybe make note that this gets a single integration point at centroid of element.

File: src/examples/mfem_common_plane.cpp [hunk 28]
```diff
@@ -78,8 +78,8 @@ int main( int argc, char** argv )
   double dt = 0.001;
   // end time of the simulation
   double t_end = 2.0;
-  // kinematic penalty parameter
-  double p_kine = 10000.0;
+  // use const penalty (true) or element penalty (false)
```
Comment (srwopschall) [inline src/examples/mfem_common_plane.cpp:81]: Does this comment make sense when you set use_const_penalty = false;?

File: src/examples/mfem_common_plane.cpp [hunk 29]
```diff
@@ -88,6 +88,8 @@ int main( int argc, char** argv )
   double lambda = 100000.0;
   // Lame parameter mu (shear modulus)
   double mu = 100000.0;
+  // kinematic penalty parameter (only for constant penalty)
```
Comment (srwopschall) [inline src/examples/mfem_common_plane.cpp:91]: Since this is an example, a host-code may want to replicate this calculation as a 'guide' for what a constant penalty would be. Can you add in the comment what this calculation is?

File: src/tests/tribol_mfem_common_plane.cpp [hunk 30]
```diff
@@ -45,21 +45,21 @@ class MfemCommonPlaneTest : public testing::TestWithParam<int> {
...
-    // kinematic penalty
-    double p_kine = 10000.0;
     // material density
     double rho = 1000.0;
     // lame parameter
     double lambda = 100000.0;
     // lame parameter (shear modulus)
     double mu = 100000.0;
+    // kinematic penalty
+    double p_kine = (lambda + 2.0 / 3.0 * mu) * std::pow(2.0, ref_levels - 1);
```
Comment (srwopschall) [inline src/tests/tribol_mfem_common_plane.cpp:64]: Just state in the comment what this calculation is.

File: src/tests/tribol_mfem_common_plane.cpp [hunk 31]
```diff
@@ -238,12 +241,13 @@ class MfemCommonPlaneTest : public testing::TestWithParam<int> {
 
 TEST_P(MfemCommonPlaneTest, mass_matrix_transfer)
 {
-  EXPECT_LT(std::abs(max_disp_ - 0.0000033982222612802126), 1.0e-8);
+  EXPECT_LT(std::abs(max_disp_ - 0.013637427890739103), 1.0e-8);
```
Comment (srwopschall) [inline src/tests/tribol_mfem_common_plane.cpp:246]: Just so I know, why did the value change so much? Different penalty?

**Review by srwopschall — COMMENTED** [review]
@ebchin - Looks great. Make consistency modifications per our review discussion. Can you report back about behavior with on-rank zero element meshes just to test out some of your checks?

File: src/examples/mfem_common_plane.cpp (hunk 28 above)
Comment (ebchin) [reply src/examples/mfem_common_plane.cpp:81]: I think so.  The example would use element penalty in that case

File: src/tests/tribol_mfem_common_plane.cpp (hunk 31 above)
Comment (ebchin) [reply src/tests/tribol_mfem_common_plane.cpp:246]: Yeah, the new penalty is much higher so there's a lot more separation at the end of the sim.

**Comment by ebchin** [comment]
Thanks @srwopschall, should be ready to go now

File: src/examples/mfem_common_plane.cpp (hunk 28 above)
Comment (srwopschall) [reply src/examples/mfem_common_plane.cpp:81]: change to (if true) and (if false)

File: src/examples/mfem_common_plane.cpp (hunk 29 above)
Comment (srwopschall) [reply src/examples/mfem_common_plane.cpp:91]: Is this calculation bulk modulus in terms of lambda and mu. Please note.

File: src/tests/tribol_mfem_common_plane.cpp (hunk 30 above)
Comment (srwopschall) [reply src/tests/tribol_mfem_common_plane.cpp:64]: Note that this is meant to be consistent with element penalty calc.

File: src/tests/tribol_mfem_common_plane.cpp [hunk 32]
```diff
@@ -36,7 +36,7 @@
  * scheme.
  *
  */
-class MfemCommonPlaneTest : public testing::TestWithParam<int> {
+class MfemCommonPlaneTest : public testing::TestWithParam<std::pair<int, tribol::KinematicPenaltyCalculation>> {
```
Comment (srwopschall) [inline src/tests/tribol_mfem_common_plane.cpp:41]: Can we add a zero refinement test on three ranks so we get a zero-element mesh on one of the ranks just as a sanity check?

**Review by srwopschall — APPROVED** [review]
@ebchin Looks good. Add those zeroeth sanity tests.

File: src/examples/mfem_common_plane.cpp [hunk 33]
```diff
@@ -344,14 +350,21 @@ int main( int argc, char** argv )
   // for computing the maximum common plane timestep and, if activated, gap rate
   // penalty.
   tribol::registerMfemVelocity(coupling_scheme_id, velocity);
-  // The type of penalty enforcement is set here (i.e. rate vs. kinematic and
-  // the method of setting the penalty value).
-  tribol::setPenaltyOptions(
```
Comment (srwopschall) [inline src/examples/mfem_common_plane.cpp:349]: Is this no longer needed? Is this call made elsewhere?

File: src/tests/tribol_mfem_common_plane.cpp [hunk 34]
```diff
@@ -199,21 +199,24 @@ class MfemCommonPlaneTest : public testing::TestWithParam<int> {
       tribol::BINNING_GRID
     );
     tribol::registerMfemVelocity(0, v);
-    tribol::setPenaltyOptions(
```
Comment (srwopschall) [inline src/tests/tribol_mfem_common_plane.cpp:202]: Why is this removed?

**Review by srwopschall — APPROVED** [review]
@ebchin - this looks great. One last thing, were you going to introduce a standalone test for the element thickness calculation (e.g. unit test), or is that contained in the tribol_mfem_common_plane test? (I may have passed over it).

File: src/examples/mfem_common_plane.cpp (hunk 33 above)
Comment (ebchin) [reply src/examples/mfem_common_plane.cpp:349]: It's done in `setMfemKinematicConstantPenalty()`

File: src/tests/tribol_mfem_common_plane.cpp (hunk 34 above)
Comment (ebchin) [reply src/tests/tribol_mfem_common_plane.cpp:202]: It gets set in `setMfemKinematicConstantPenalty()` and `setMfemKinematicElementPenalty()`

**Comment by ebchin** [comment]
> @ebchin - this looks great. One last thing, were you going to introduce a standalone test for the element thickness calculation (e.g. unit test), or is that contained in the tribol_mfem_common_plane test? (I may have passed over it).

Yeah, `tribol_mfem_common_plane` tests the element thickness calculation and compares it to a constant penalty tuned to the element thickness calculation.
